File: src/lexer/Literal.ts
import {Terminal} from '../common/Terminal';
import {Automaton, AutomatonBuilder, State} from './automaton/Automaton';
import { CharConstraint } from './CharConstraint';

/**
 * A Literal is a fixed sequence of characters (i.e a keyword)
 */
export class Literal extends Terminal {
    
    chars:string;

    constructor(chars:string) {
        super("Literal('" + chars + "')");
        this.chars = chars;
    }

    get automaton():Automaton {
        let builder = AutomatonBuilder.forTokenType(this);
        let currentState = builder.initialState();
        for(let i=0; i < this.chars.length; i+=1) {
            let c = this.chars[i];
            let nextState:State = i == this.chars.length-1 ? builder.newFinalState() : builder.newNonFinalState();
            let transition = currentState.addTransition(CharConstraint.eq(c));
            transition.target = nextState;
            currentState = nextState;
        }
        return builder.build();
    }

};

File: old/lexer/src/main/java/net/jr/lexer/basicterminals/Literal.java
package net.jr.lexer.basicterminals;

import net.jr.lexer.automaton.Automaton;
import net.jr.lexer.automaton.DefaultAutomaton;
import net.jr.lexer.impl.TerminalImpl;

import java.io.DataOutput;
import java.io.IOException;

import static net.jr.lexer.impl.CharConstraint.Builder.eq;

/**
 * A literal is a 'keyword' i.e a fixed string.
 */
public class Literal extends TerminalImpl {

    private String value;

    public Literal(String value) {
        this();
        this.value = value;
    }

    private Literal() {
        setPriority(2);
    }

    public static Literal unMarshall(java.io.DataInput in) throws IOException {
        Literal l = TerminalImpl.unMarshall(new Literal(), in);
        l.value = in.readUTF();
        return l;
    }

    @Override
    public Automaton getAutomaton() {
        DefaultAutomaton.Builder builder = DefaultAutomaton.Builder.forTokenType(this);
        DefaultAutomaton.Builder.BuilderState currentState = builder.initialState();
        char[] chars = value.toCharArray();
        for (int i = 0, max = chars.length; i < max; i++) {
            char c = chars[i];
            final DefaultAutomaton.Builder.BuilderState targetState;
            if (i == max - 1) {
                targetState = builder.newFinalState();
            } else {
                targetState = builder.newNonFinalState();
            }
            currentState.when(eq(c)).goTo(targetState);
            currentState = targetState;
        }
        return builder.build();
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null) {
            return false;
        }
        if (!obj.getClass().equals(Literal.class)) {
            return false;
        }
        final Literal o = (Literal) obj;
        return value.equals(o.value);
    }

    @Override
    public int hashCode() {
        return 23 + value.hashCode();
    }

    @Override
    public String toString() {
        String name = getName();
        return name == null ? "'" + value + "'" : name;
    }

    @Override
    public void marshall(DataOutput dataOutputStream) throws IOException {
        super.marshall(dataOutputStream);
        dataOutputStream.writeUTF(value);
    }
}
